File: platform/graphs/workflow_dsl.py
"""workflow DSL — source of truth for the designer.

Document shape matches workflow JSON:
  identifier, title, description,
  nodes[{identifier, title, icon?, description?, config, ...}],
  connections[{sourceIdentifier, targetIdentifier,
               sourceOutletIdentifier? | sourceOptionIdentifier?, fallback?}]

There is no End node — a run finishes when leaf nodes have no outgoing
connections.

UI layout (x/y) is NOT part of workflow DSL — it lives in optional ``ui`` and is
stripped by ``normalize`` before Temporal parse.
"""

from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, Field

# config.type values (no END)
WORKFLOW_CONFIG_TYPES = {
    "SELF_SERVE_TRIGGER",
    "EVENT_TRIGGER",
    "SCHEDULE_TRIGGER",
    "WEBHOOK",
    "KAFKA",
    "UPSERT_ENTITY",
    "INTEGRATION_ACTION",
    "INTERNAL_SERVICE",
    "AI",
    "AI_AGENT",
    "CONDITION",
    "INPUT",
    "SUBFLOW",
}

_CANVAS_TYPE = {
    "SELF_SERVE_TRIGGER": "trigger",
    "EVENT_TRIGGER": "trigger",
    "SCHEDULE_TRIGGER": "trigger",
    "AI_AGENT": "ai_agent",
    "AI": "ai",
    "WEBHOOK": "webhook",
    "CONDITION": "condition",
    "INPUT": "input",
    "UPSERT_ENTITY": "upsert_entity",
    "KAFKA": "kafka",
    "INTEGRATION_ACTION": "integration_action",
    "INTERNAL_SERVICE": "internal_service",
    "SUBFLOW": "subflow",
}

_CANVAS_TO_CONFIG = {
    "trigger": "SELF_SERVE_TRIGGER",
    "ai_agent": "AI_AGENT",
    "ai": "AI",
    "webhook": "WEBHOOK",
    "condition": "CONDITION",
    "input": "INPUT",
    "upsert_entity": "UPSERT_ENTITY",
    "kafka": "KAFKA",
    "integration_action": "INTEGRATION_ACTION",
    "internal_service": "INTERNAL_SERVICE",
    "subflow": "SUBFLOW",
    # end intentionally omitted — dropped
}


class DslConnection(BaseModel):
    sourceIdentifier: str
    targetIdentifier: str
    sourceOutletIdentifier: Optional[str] = None  # INPUT outlets
    sourceOptionIdentifier: Optional[str] = None  # CONDITION options
    fallback: bool = False


class DslNode(BaseModel):
    identifier: str
    title: str = ""
    icon: Optional[str] = None
    description: Optional[str] = None
    config: dict[str, Any] = Field(default_factory=dict)
    variables: Optional[dict[str, Any]] = None
    links: Optional[list[str]] = None
    verbose: Optional[bool] = None


class WorkflowDsl(BaseModel):
    """workflow document."""

    identifier: str
    title: str = ""
    description: str = ""
    produces: list[str] = Field(default_factory=list)
    nodes: list[DslNode] = Field(default_factory=list)
    connections: list[DslConnection] = Field(default_factory=list)
    ui: Optional[dict[str, Any]] = None

    def trigger_node(self) -> DslNode | None:
        for n in self.nodes:
            cfg_type = str((n.config or {}).get("type") or "").upper()
            if cfg_type.endswith("TRIGGER") or cfg_type in {
                "SELF_SERVE_TRIGGER",
                "EVENT_TRIGGER",
                "SCHEDULE_TRIGGER",
            }:
                return n
        return self.nodes[0] if self.nodes else None


def parse_workflow_dsl(raw: dict[str, Any] | str) -> WorkflowDsl:
    """Parse workflow DSL from dict or JSON string."""
    import json

    if isinstance(raw, str):
        data = json.loads(raw)
    else:
        data = raw
    if not isinstance(data, dict):
        raise ValueError("workflow DSL must be an object")
    return WorkflowDsl.model_validate(data)


def _edge_layout_key(source: str, target: str, handle: str | None) -> str:
    h = handle if handle and handle != "out" else ""
    return f"{source}::{target}::{h}"


def workflow_dsl_to_canvas(doc: WorkflowDsl) -> dict[str, Any]:
    """Map workflow DSL → React Flow-friendly structure (designer only)."""
    ui = doc.ui or {}
    positions = ui.get("positions") or {}
    waypoints = ui.get("edgeWaypoints") or {}
    nodes = []
    for n in doc.nodes:
        cfg_type = str((n.config or {}).get("type") or "").upper()
        if cfg_type == "END":
            continue
        cfg = dict(n.config or {})
        meta = {}
        if cfg.get("governance") or cfg.get("immutable") or str(n.identifier).startswith("governance-"):
            meta = {"immutable": True, "governance": True}
        pos = positions.get(n.identifier) if isinstance(positions, dict) else None
        nodes.append({
            "id": n.identifier,
            "type": _CANVAS_TYPE.get(cfg_type, cfg_type.lower() or "webhook"),
            "label": n.title or n.identifier,
            "position": pos if isinstance(pos, dict) else {"x": 0, "y": 0},
            "config": cfg,
            "metadata": meta,
        })
    keep = {n["id"] for n in nodes}
    edges = []
    for i, c in enumerate(doc.connections):
        if c.sourceIdentifier not in keep or c.targetIdentifier not in keep:
            continue
        outlet = c.sourceOptionIdentifier or c.sourceOutletIdentifier
        wp = waypoints.get(_edge_layout_key(c.sourceIdentifier, c.targetIdentifier, outlet)) if isinstance(waypoints, dict) else None
        edges.append({
            "id": f"c{i}-{c.sourceIdentifier}-{c.targetIdentifier}",
            "source": c.sourceIdentifier,
            "target": c.targetIdentifier,
            "sourceHandle": outlet,
            "condition": outlet,
            "waypoint": wp if isinstance(wp, dict) else None,
        })
    return {
        "graphId": doc.identifier,
        "name": doc.title,
        "description": doc.description,
        "kind": ((doc.ui or {}).get("kind") if doc.ui else None) or "e2e",
        "produces": list(doc.produces or []),
        "nodes": nodes,
        "edges": edges,
        "entrypoints": [doc.trigger_node().identifier] if doc.trigger_node() else [],
    }

File: frontend/src/workflowDsl.ts
/** Workflow DSL ↔ React Flow canvas helpers.

Workflows have no End node — leaf actions/conditions terminate the run.
CONDITION branches use ``options`` + connection ``sourceOptionIdentifier``.
INPUT uses ``outlets`` + ``sourceOutletIdentifier``.
*/

import type { GraphKind, WorkflowEdge, WorkflowGraph, WorkflowNode, NodeType } from "./types";
import { isGovernanceNode, readProduces } from "./governance";

const CANVAS_TO_CONFIG: Record<string, string> = {
  trigger: "SELF_SERVE_TRIGGER",
  ai_agent: "AI_AGENT",
  ai: "AI",
  webhook: "WEBHOOK",
  condition: "CONDITION",
  input: "INPUT",
  upsert_entity: "UPSERT_ENTITY",
  kafka: "KAFKA",
  integration_action: "INTEGRATION_ACTION",
  internal_service: "INTERNAL_SERVICE",
  subflow: "SUBFLOW",
};

const CONFIG_TO_CANVAS: Record<string, NodeType> = {
  SELF_SERVE_TRIGGER: "trigger",
  EVENT_TRIGGER: "trigger",
  SCHEDULE_TRIGGER: "trigger",
  AI_AGENT: "ai_agent",
  AI: "ai",
  WEBHOOK: "webhook",
  CONDITION: "condition",
  INPUT: "input",
  UPSERT_ENTITY: "upsert_entity",
  KAFKA: "kafka",
  INTEGRATION_ACTION: "integration_action",
  INTERNAL_SERVICE: "internal_service",
  SUBFLOW: "subflow",
};

export type WorkflowDsl = {
  identifier: string;
  title: string;
  description?: string;
  nodes: Array<{
    identifier: string;
    title: string;
    description?: string;
    config: Record<string, unknown>;
  }>;
  connections: Array<{
    sourceIdentifier: string;
    targetIdentifier: string;
    sourceOutletIdentifier?: string | null;
    sourceOptionIdentifier?: string | null;
    fallback?: boolean;
  }>;
  ui?: {
    positions?: Record<string, { x: number; y: number }>;
    kind?: GraphKind;
    edgeWaypoints?: Record<string, { x: number; y: number }>;
  };
  produces?: string[];
};

export function isWorkflowDsl(doc: unknown): doc is WorkflowDsl {
  if (!doc || typeof doc !== "object") return false;
  const value = doc as Record<string, unknown>;
  return (
    typeof value.identifier === "string" &&
    Array.isArray(value.nodes) &&
    Array.isArray(value.connections) &&
    !Array.isArray(value.edges)
  );
}

export function isCanvasWorkflowGraph(doc: unknown): doc is WorkflowGraph {
  if (!doc || typeof doc !== "object") return false;
  const value = doc as Record<string, unknown>;
  return (
    Array.isArray(value.nodes) &&
    Array.isArray(value.edges) &&
    value.nodes.every((node) => !!node && typeof node === "object" && typeof (node as { id?: unknown }).id === "string")
  );
}

/** Accept the canonical workflow DSL and the API's canvas projection. */
export function workflowDocumentToCanvas(
  document: unknown,
  kind: GraphKind = "e2e",
): WorkflowGraph | null {
  let value = document;
  if (typeof value === "string") {
    try {
      value = JSON.parse(value);
    } catch {
      return null;
    }
  }
  if (isWorkflowDsl(value)) return workflowDslToCanvas(value, kind);
  if (!isCanvasWorkflowGraph(value)) return null;
  return {
    ...value,
    graphId: value.graphId || "",
    version: value.version || "0.1.0",
    name: value.name || "",
    kind: value.kind || kind,
    entrypoints: value.entrypoints || [],
    nodes: value.nodes,
    edges: value.edges,
  };
}

export function edgeLayoutKey(source: string, target: string, handle?: string | null) {
  const h = handle && handle !== "out" ? handle : "";
  return `${source}::${target}::${h}`;
}

export function canvasToWorkflowDsl(graph: WorkflowGraph): WorkflowDsl {
  const positions: Record<string, { x: number; y: number }> = {};
  const byId = new Map((graph.nodes || []).map((n) => [n.id, n]));
  const nodes = (graph.nodes || [])
    .filter((n) => n.type !== ("end" as NodeType) && String(n.config?.type || "").toUpperCase() !== "END")
    .map((n) => {
      if (n.position && Number.isFinite(Number(n.position.x)) && Number.isFinite(Number(n.position.y))) {
        positions[n.id] = { x: Number(n.position.x), y: Number(n.position.y) };
      }
      const cfg = { ...(n.config || {}) };
      if (!cfg.type) {
        cfg.type =
          n.type === "trigger"
            ? (cfg.triggerType as string) || "SELF_SERVE_TRIGGER"
            : CANVAS_TO_CONFIG[n.type] || String(n.type).toUpperCase();
      }
      // CONDITION: uses options[]
      if (String(cfg.type).toUpperCase() === "CONDITION") {
        delete (cfg as { outlets?: unknown }).outlets;
        delete (cfg as { outcomes?: unknown }).outcomes;
      }
      delete (cfg as { skill_keys?: unknown }).skill_keys;
      delete (cfg as { skill_key?: unknown }).skill_key;
      delete (cfg as { rule_keys?: unknown }).rule_keys;
      if (String(cfg.type).toUpperCase() !== "SUBFLOW") {
        delete (cfg as { playbook_key?: unknown }).playbook_key;
        delete (cfg as { flow_key?: unknown }).flow_key;
        delete (cfg as { graph_key?: unknown }).graph_key;
        delete (cfg as { graph_id?: unknown }).graph_id;
        delete (cfg as { graph_kind?: unknown }).graph_kind;
      }
      delete (cfg as { testOutputs?: unknown }).testOutputs;
      const locked = isGovernanceNode({ id: n.id, config: cfg, metadata: n.metadata });
      if (locked) {
        cfg.governance = true;
        cfg.immutable = true;
      }
      return {
        identifier: n.id,
        title: n.label || n.id,
        config: cfg,
      };
    });
  const keep = new Set(nodes.map((n) => n.identifier));
  const edgeWaypoints: Record<string, { x: number; y: number }> = {};
  const connections = (graph.edges || [])
    .filter((e) => keep.has(e.source) && keep.has(e.target))
    .map((e) => {
      const handle = e.sourceHandle || e.condition || undefined;
      if (e.waypoint && Number.isFinite(e.waypoint.x) && Number.isFinite(e.waypoint.y)) {
        edgeWaypoints[edgeLayoutKey(e.source, e.target, handle)] = { x: e.waypoint.x, y: e.waypoint.y };
      }
      const src = byId.get(e.source);
      const srcType = String(src?.config?.type || "").toUpperCase() || (src?.type === "condition" ? "CONDITION" : src?.type === "input" ? "INPUT" : "");
      if (srcType === "CONDITION") {
        return {
          sourceIdentifier: e.source,
          targetIdentifier: e.target,
          sourceOptionIdentifier: handle || undefined,
          fallback: e.fallback ?? false,
        };
      }
      return {
        sourceIdentifier: e.source,
        targetIdentifier: e.target,
        sourceOutletIdentifier: handle || undefined,
        fallback: e.fallback ?? false,
      };
    });
  return {
    identifier: graph.graphId || "workflow",
    title: graph.name || "",
    description: graph.description || "",
    produces: readProduces(graph),
    nodes,
    connections,
    ui: { positions, kind: graph.kind, ...(Object.keys(edgeWaypoints).length ? { edgeWaypoints } : {}) },
  };
}

export function workflowDslToCanvas(dsl: WorkflowDsl, kind: GraphKind = "e2e"): WorkflowGraph {
  const positions = dsl.ui?.positions || {};
  const nodes: WorkflowNode[] = (dsl.nodes || [])
    .filter((n) => String((n.config || {}).type || "").toUpperCase() !== "END")
    .map((n) => {
      const cfgType = String((n.config || {}).type || "").toUpperCase();
      const type: NodeType = CONFIG_TO_CANVAS[cfgType] || (cfgType.toLowerCase() as NodeType);
      const config = { ...(n.config || {}) };
      const locked = isGovernanceNode({ id: n.identifier, config });
      return {
        id: n.identifier,
        type,
        label: n.title || n.identifier,
        position: positions[n.identifier] || { x: 0, y: 0 },
        config,
        metadata: locked ? { immutable: true, governance: true } : undefined,
      };
    });
  const keep = new Set(nodes.map((n) => n.id));
  const wps = dsl.ui?.edgeWaypoints || {};
  const edges: WorkflowEdge[] = (dsl.connections || [])
    .filter((c) => keep.has(c.sourceIdentifier) && keep.has(c.targetIdentifier))
    .map((c, i) => {
      const handle = c.sourceOptionIdentifier || c.sourceOutletIdentifier || undefined;
      const wp = wps[edgeLayoutKey(c.sourceIdentifier, c.targetIdentifier, handle)];
      return {
        id: `c${i}-${c.sourceIdentifier}-${c.targetIdentifier}`,
        source: c.sourceIdentifier,
        target: c.targetIdentifier,
        sourceHandle: handle,
        condition: handle,
        fallback: c.fallback,
        waypoint: wp || null,
      };
    });
  return {
    graphId: dsl.identifier,
    version: "0.1.0",
    name: dsl.title,
    description: dsl.description || "",
    kind: dsl.ui?.kind || kind,
    produces: dsl.produces || [],
    entrypoints: nodes.filter((n) => n.type === "trigger").map((n) => n.id),
    nodes,
    edges,
  };
}
